File: ros2/workspace/src/simple_ros_cpp/src/service_client.cpp
#include <iostream>
#include <chrono> 
#include "rclcpp/rclcpp.hpp"
#include "std_srvs/srv/set_bool.hpp"

using namespace std; 
void printHello(); 

typedef shared_ptr<std_srvs::srv::SetBool::Request> Req; 
typedef shared_ptr<std_srvs::srv::SetBool::Response> Res; 

class ServiceClient : public rclcpp::Node {
    public:
        ServiceClient() : Node("ServiceClient") {
            _client = create_client<std_srvs::srv::SetBool>("setBoolService"); 
            while(!_client->wait_for_service(1s) ) {
                RCLCPP_INFO(get_logger(), "service is not available!! "); 
            }    
            _request = std::make_shared<std_srvs::srv::SetBool::Request>(); 
            RCLCPP_INFO(get_logger(), "service is available!! "); 
            _request->data = isTrue; 
            send_request(_request); 

            // _twist_pub_timer = create_wall_timer(1s, std::bind(&ServiceClient::twist_pub, this)); 
             _update_timer = create_wall_timer(1s, std::bind(&ServiceClient::update, this)); 
        }
    private:
        int _count; 
        rclcpp::Client<std_srvs::srv::SetBool>::SharedPtr _client; 
        rclcpp::TimerBase::SharedPtr _twist_pub_timer; 
        rclcpp::TimerBase::SharedPtr _update_timer; 
        bool isTrue = false; 
        Req _request;  

        void send_request(Req request) {
            RCLCPP_INFO(get_logger(), "send_request~~ "); 
            auto future = _client->async_send_request(request, bind(&ServiceClient::receive_callback, this, placeholders::_1)); 
        }

        void receive_callback(rclcpp::Client<std_srvs::srv::SetBool>::SharedFuture future) {
            Res response = future.get(); 
            // string suc = (response->success)?"true":"false"; 
            RCLCPP_INFO(get_logger(), "result : %s", (response->success)?"true":"false"); 
            RCLCPP_INFO(get_logger(), "message : %s", response->message.c_str()); 
        }
        
        void update() {
            _request->data = isTrue = !isTrue; 
            send_request(_request); 
        }
}; 

int main()
{
    rclcpp::init(0, nullptr);
    auto node = std::make_shared<ServiceClient>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2/workspace/src/simple_ros/simple_ros/service_client.py
import rclpy
from rclpy.node import Node
from rclpy.qos import (
    QoSDurabilityPolicy,
    QoSHistoryPolicy,
    QoSProfile,
    QoSReliabilityPolicy,
)
from std_msgs.msg import String
from std_srvs.srv import SetBool 

class Service_client(Node):
    def __init__(self):
        super().__init__("Service_client")
        self.create_timer(0.5, self.print_status)
        self.client = self.create_client(SetBool, "setBoolService") 
        while not self.client.wait_for_service(timeout_sec=10.0):
            self.get_logger().info("service is not available!!")

        self.request = SetBool.Request()

        self.create_timer(2.0, self.send_request)
        # self.send_request()   
        self.bool = bool()
    
    def send_request(self):
        self.bool = not self.bool 
        self.request.data = self.bool
        self.future = self.client.call_async(self.request)
        self.future.add_done_callback(self.response_callback)

    def response_callback(self, future):
        response : SetBool.Response = future.result()
        self.get_logger().info(f"result : {response.success}")
        self.get_logger().info(f"message : {response.message}")

    def setBool_callback(self, request:SetBool.Request, response:SetBool.Response):
        self.get_logger().info(f"request data : {request.data}")
        self.bool = request.data

        if request.data:
            response.success = True
            response.message = "True response"
        else :
            response.success = False
            response.message = "False response"

        self.get_logger().info(response.message)
        return response

    def print_status(self):
        self.get_logger().info(f"current bool : {self.bool}")

def main():
    rclpy.init()
    node = Service_client()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info(f"KeyboardInterrupted")
    finally:
        node.destroy_node()

if __name__ == "__main__":
    main()
